Merging new model zoo into existing onnx/models

abhinavs95 opened the pull request:
This PR merges the new model zoo(ONNX/model-zoo) into the existing one(ONNX/models).

It addresses following issues too:

* Reorganization of main README for separate model classes
* Add synthetic input/output datasets to each model
* Remove all sections related to Model Server and links to Model Server Archives . Can add link to some other location
* Remove preprocessing/postprocessing code from model README and add links to point to a file containing it.
* Add details of ONNX version with which model was created.
* Add licensing information for each model.
* Update model description to include a header that is less technical, includes model use cases.

In #74 the commit history from the new model zoo was lost so re-raising this PR and closing old one.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/onnx/models?pullRequest=75) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/onnx/models?pullRequest=75) before we can accept your contribution.<br/>**1** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: ankkhedia<br/>:x: abhinavs95<br/>:x: ThomasDelteil<br/>:x: lupesko<br/>:x: rahulbhalley<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/onnx/models?pullRequest=75) it.</sub>

prasanthpul commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,66 @@
+Contributors' name
+<hr>
+
+## Acknowledgements
```
I recommend combining Contributors and Acknowlegements

abhinavs95 replied:
I feel combining Acknowledgements with References makes more sense because tags under acknowledgements is used by the contributor to help in the contribution of the model, just like content under references. What do you think @lupesko ?

lupesko replied:
Agree with @abhinavs95 
This is the template for contributing models, and there is a difference between listing yourself as the contributor of the model, to the list of references/acks you list out (e.g. paper, reference implementations, etc.)
Suggest we simplify per @abhinavs95 's suggestion

prasanthpul replied:
That's fine.

abhinavs95 replied:
merged Acknowledgements with References

prasanthpul commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,66 @@
+Thank anyone whose existing work was used
+<hr>
+
+## Keywords
```
Let's remove Keywords. The full text will be indexed for search purposes anyways and the keywords don't add value, only add opportunity for mischief.

abhinavs95 replied:
removed keywords from all model readme pages

prasanthpul commented on contribute.md:
```diff
@@ -0,0 +1,5 @@
+# How to Contribute to the Model Zoo
```
Need to add info that any model submitted must pass ONNX model checker. (it's manually run for now until we setup CI)

abhinavs95 replied:
Added model checker info in PR template with code on how to perform the check.. Thanks!

jspisak replied:
@rahulbhalley - Can you sign the CLA?

ThomasDelteil commented on models/image_classification/imagenet_postprocess.py:
```diff
@@ -0,0 +1,13 @@
+def postprocess(scores): 
+    '''
+    Postprocessing with mxnet gluon
+    The function takes scores generated by network and returns the class IDs in decreasing order
```
by the*

abhinavs95 replied:
fixed, thanks!

ThomasDelteil commented on models/image_classification/imagenet_inference.ipynb:
```diff
@@ -0,0 +1,293 @@
+    }
+   ],
+   "source": [
+    "os.system('curl -O https://s3.amazonaws.com/model-server/inputs/kitten.jpg')"
```
use `mxnet.test_utils.download()`

abhinavs95 replied:
done

ThomasDelteil commented on models/image_classification/imagenet_inference.ipynb:
```diff
@@ -0,0 +1,293 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "os.system('curl -O https://s3.amazonaws.com/onnx-model-zoo/synset.txt')\n",
```
use `mxnet.test_utils.download()`

abhinavs95 replied:
done

ThomasDelteil commented on models/image_classification/mobilenet/README.md:
```diff
@@ -0,0 +1,68 @@
+# MobileNet
+
+## Use cases
+Mobilenet models perform image classification - they take images as input and classify the major object in the image into a set of pre-defined classes. They are trained on ImageNet dataset which contains images from 1000 classes. MobileNet models are also very efficient in terms of speed and size and hence are ideal for embedded and mobile applications.
```
MobileNet

abhinavs95 replied:
fixed.. thanks!

ankkhedia commented:
@prasanthpul @lupesko All the issues mentioned have been addressed. We are waiting for reply from you guys on one comment.

prasanthpul commented:
@mx-iao please review to ensure all your items are covered

mx-iao commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,68 @@
+
+## Model
+
+ |Model        |Download  | Download (with sample test data)|ONNX version|Accuracy |
```
Can you add a column `MD5 checksum` in the table for the checksum of the model, so users can check to see if a model file has been changed/updated since they last downloaded it

mx-iao replied:
And an `Opset version` column for the model

abhinavs95 replied:
done!

mx-iao commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,68 @@
+<hr>
+
+## Training
+Training details (preprocessing, hyperparameters, resources and environment) along with link to a training notebook (optional)
```
Should also clarify that if model was not trained from scratch but instead sourced from elsewhere (e.g. converted from a CoreML model), the contributor should also explain what the process/source was for obtaining the final ONNX model

abhinavs95 replied:
done!

mx-iao commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,68 @@
+### Input to model
+Input to network (Example: 224x224 pixels in RGB)
+
+### Preprocessing steps
```
Add that  framework- and platform-agnostic code sample is preferred here, or pseudocode if that's not possible

mx-iao commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,68 @@
+### Output of model
+Output of network
+
+### Postprocessing steps
```
same comment as for Preprocessing steps

mx-iao commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,68 @@
+
+ |Model        |Download  | Download (with sample test data)|ONNX version|Accuracy |
+|-------------|:--------------|:--------------|:--------------|:--------------|
+|Model Name       |ONNX Model download link with size  |  tar file containing ONNX model and synthetic test data|ONNX version used for conversion|Accuracy values |
```
Should we clarify here that the test data should be in the `.pb` format

abhinavs95 replied:
updated PR template with clarification

houseroad commented on PULL_REQUEST_TEMPLATE.md:
```diff
@@ -0,0 +1,68 @@
+import onnx
+from onnx import checker
+model_proto = onnx.load("path to .onnx file")
+checker.check_graph(model_proto.graph)
```
Please use `checker.check_model(model_proto)`

abhinavs95 replied:
updated.. thanks!

houseroad commented on README.md:
```diff
@@ -1,40 +1,159 @@
 
-### Models
+<p align="center">
+	<img src="images/ONNX Model Zoo Graphics.png" width="60%"/>
```
please remove the empty space in the file name, thanks

abhinavs95 replied:
updated to ONNX_Model_Zoo_Graphics.png

houseroad commented on models/image_classification/imagenet_postprocess.py:
```diff
@@ -0,0 +1,13 @@
+import mxnet as mx
```
`import numpy as np`

houseroad replied:
the only use of `mx` is the softmax function.

You can implement the `softmax` like this

```
def softmax(X, theta = 1.0, axis = None):
    """
    Compute the softmax of each element along an axis of X.

    Parameters
    ----------
    X: ND-Array. Probably should be floats. 
    theta (optional): float parameter, used as a multiplier
        prior to exponentiation. Default = 1.0
    axis (optional): axis to compute values along. Default is the 
        first non-singleton axis.

    Returns an array the same size as X. The result will sum to 1
    along the specified axis.
    """

    # make X at least 2d
    y = np.atleast_2d(X)

    # find axis
    if axis is None:
        axis = next(j[0] for j in enumerate(y.shape) if j[1] > 1)

    # multiply y against the theta parameter, 
    y = y * float(theta)

    # subtract the max for numerical stability
    y = y - np.expand_dims(np.max(y, axis = axis), axis)
    
    # exponentiate y
    y = np.exp(y)

    # take the sum along the specified axis
    ax_sum = np.expand_dims(np.sum(y, axis = axis), axis)

    # finally: divide elementwise
    p = y / ax_sum

    # flatten if X was 1D
    if len(X.shape) == 1: p = p.flatten()

    return p
```

reference: https://nolanbconaway.github.io/blog/2017/softmax-numpy

lupesko replied:
@houseroad we all know what softmax is, and how to implement it.
However, I think using a library function makes a better case for readability and reducing the cognitive load on users, so I don't think it makes sense to implement softmax here.
As a side note, the issue of using framework specific code (in this case MXNet) for pre/post processing in a separate file was already discussed and approved in a review meeting.

houseroad replied:
Yes, people know `softmax`. My point is that if we can get rid of the dependency of specific backend, user experience can be improved a lot. Not all the users would like to install another backend just for preprocessing and postprocessing. If they don't install mxnet, they have write the preprocessing and postprocessing script by themselves. Such changes for us are so easy to make, that's why I suggest to create simple libraries using numpy.

jspisak replied:
@lupesko - I think this goes back to our discussion around neutrality and using OpenCV/Numpy for generic pre/post processing. I think it is OK to link to mxnet examples but the model zoo, as we discussed, should not be any more tied to a particular platform than required. Sound reasonable?

lupesko replied:
The agreement was to not have framework specific code in the model readme, and instead link framework specific file(s). I think this is exactly what the contributor has done here.

Looks like you are asking to host this python code outside of the repo? if we do that, the user experience will be exactly the same (link to a code file), but maintenance and review will be more difficult (since we will have cross repos dependencies). Please consider this point...

houseroad commented on models/image_classification/imagenet_postprocess.py:
```diff
@@ -0,0 +1,13 @@
+    prob = mx.ndarray.softmax(scores).asnumpy()
+    prob = np.squeeze(prob)
+    a = np.argsort(prob)[::-1]
+    return a
```
newline at the end please

houseroad commented on models/image_classification/imagenet_preprocess.py:
```diff
@@ -0,0 +1,19 @@
+    Preprocessing required on the images for inference with mxnet gluon
+    The function takes path to an image and returns processed tensor
+    '''
+    transform_fn = transforms.Compose([
```
similarly, could you use numpy to implement these transformation functions?

It won't be very hard.

lupesko replied:
@houseroad same as my comment above.

houseroad commented on models/image_classification/mobilenet/README.md:
Could you also specify the opset version here? working with which version of onnx?

mx-iao replied:
We can add an `Opset version` column to the model table in PULL_REQUEST_TEMPLATE.md

abhinavs95 replied:
added opset version for the models and also in PR template

houseroad commented on models/image_classification/resnet/README.md:
```diff
@@ -0,0 +1,84 @@
+ResNet v2 uses pre-activation function whereas ResNet v1  uses post-activation for the residual blocks. The models below have 18, 34, 50, 101 and 152 layers for with ResNetv1 and ResNetv2 architecture.
+
+* Version 1:
+
```
Could you also specify the opset version here? working with which version of onnx?

abhinavs95 replied:
done!

houseroad commented on models/image_classification/squeezenet/README.md:
```diff
@@ -0,0 +1,59 @@
+Squeezenet 1.1 gives AlexNet level of accuracy with 50X fewer parameters.
+<!-- SqueezeNet 1.1 has 2.4x less computation and slightly fewer parameters than SqueezeNet 1.0, without sacrificing accuracy.-->
+
+|Model        |Download  | Download (with sample test data)|ONNX Version|Top-1 accuracy (%)|Top-5 accuracy (%)|
```
Could you also specify the opset version here? working with which version of onnx?

abhinavs95 replied:
done!

houseroad commented on models/image_classification/vgg/README.md:
```diff
@@ -0,0 +1,73 @@
+The models below are variant of same network with different number of layers and use of batch normalisation. VGG 16 and VGG 19 have 16 and 19 convolutional layers respectively. VGG 16_bn and VGG 19_bn have the same architecture as their original counterparts but with batch normalization applied after each convolutional layer, which leads to better convergence and slightly better accuracies.
+
+ |Model        |Download  | Download (with sample test data)|ONNX Version|Top-1 accuracy (%)|Top-5 accuracy (%)|
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|
```
Could you also specify the opset version here? working with which version of onnx?

abhinavs95 replied:
done!

abhinavs95 commented:
@mx-iao @houseroad @prasanthpul thanks for the review!

The following changes have been made acc to your feedback:
* Added MD5 checksum for each model (link to a text file containing the string)
* Added opset version for each model
* Updated PR template
* Fixed other errors/typos that were pointed out

lupesko commented:
We'd really want to get this merged.

@aaronmarkham can you please review?
@prasanthpul your comments were all addressed I believe... can you please approve the request?
@jspisak and @houseroad hope all your feedback was addressed? would be great to get your approval

Thanks all!
